File: 138. Copy List with Random Pointer.py
# Problem Link: https://leetcode.com/problems/copy-list-with-random-pointer/

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return head

        node_map = {}
        temp = head
        while temp:
            node_map[temp] = Node(temp.val)
            temp = temp.next

        temp, nxt = head, head.next
        while temp:
            if nxt:
                node_map[temp].next = node_map[nxt]
            temp = nxt
            if nxt:
                nxt = nxt.next

        temp = head
        while temp:
            if temp.random:
                node_map[temp].random = node_map[temp.random]
            temp = temp.next

        return node_map[head]

File: C#/138. Copy List with Random Pointer.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;
    
    public Node(int _val) {
        val = _val;
        next = null;
        random = null;
    }
}
*/

public class Solution {
    public Node CopyRandomList(Node head) {
        if (head == null){
            return head;
        }
        
        Dictionary<Node, Node> nodeMap = new Dictionary<Node, Node>();
        
        var temp = head;
        
        while(temp != null){
            nodeMap[temp] = new Node(temp.val);
            temp = temp.next;
        }
        
        temp = head;
        
        while(temp != null){
            if (temp.next != null){
                nodeMap[temp].next = nodeMap[temp.next];
            }
            temp = temp.next;
        }
        
        temp = head;
        
        while(temp != null){
            if(temp.random != null){
                nodeMap[temp].random = nodeMap[temp.random];
            }
            temp = temp.next;
        }
        
        return nodeMap[head];
    }
}